### PR TITLE
Fix Minimax M2 partial_rotary_factor by remapping legacy rotary_dim

### LINKED ISSUE #48241: MiniMaxM2 silently applies full-head RoPE: config drops the checkpoints' legacy `rotary_dim` field
### System Info

Verified on current `main` (`3d6c720188`, `5.16.0.dev0`); also present in v5.12.1. torch 2.10, linux.

### Reproduction

All released `MiniMaxAI/MiniMax-M2*` checkpoints are trained with **partial RoPE** (`rotary_dim: 64` of `head_dim: 128`) and express it in `config.json` through a legacy `rotary_dim` field — their original `trust_remote_code` config class derived `partial_rotary_factor = rotary_dim / head_dim` in `__init__`. The in-tree `MiniMaxM2Config` keeps `rotary_dim` as an opaque extra attribute and `RopeParametersMixin.convert_rope_params_to_dict` only recognizes the standard `partial_rotary_factor` spelling, so `rope_parameters` never receives a partial factor and `MiniMaxM2RotaryEmbedding` silently rotates **all 128 head dims with a full-width frequency ladder** — both the frequency spacing and the rotate-half pairing are wrong in every attention layer, with no warning.

Config-level repro (no weights needed):

```python
from transformers import MiniMaxM2Config
from transformers.models.minimax_m2.modeling_minimax_m2 import MiniMaxM2RotaryEmbedding

# Mirrors MiniMaxAI/MiniMax-M2* config.json (rotary_dim: 64, head_dim: 128).
config = MiniMaxM2Config(rotary_dim=64)
print(config.rope_parameters)   # {'rope_theta': 5000000.0, 'rope_type': 'default'}  <- no partial factor
print(MiniMaxM2RotaryEmbedding(config).inv_freq.shape)
# torch.Size([64]) -> full-head rotation. Expected torch.Size([32]) for rotary_dim=64.
```

Or directly from the released checkpoint's metadata:

```python
from transformers import AutoConfig
config = AutoConfig.from_pretrained("MiniMaxAI/MiniMax-M2")
print(config.rotary_dim, config.head_dim, config.rope_parameters)
# 64 128 {'rope_theta': 5000000, 'rope_type': 'default'}
```

### Measured impact

We hit this in NVIDIA NeMo AutoModel's checkpoint-parity CI while using the in-tree implementation as the vanilla reference for `MiniMaxAI/MiniMax-M2.7`:

- Full-model logits vs a faithful implementation of the checkpoint's own modeling semantics on a 2,048-token document: **mean KL 13.1, p95 22.4, cosine 0.02** — the model is effectively destroyed while loading cleanly.
- Minimal causal proof: on a tiny randomly-initialized 2-layer model in fp32, the in-tree forward diverges from the reference (cosine 0.885); adding `"partial_rotary_factor": 0.5` to `rope_parameters` on the same config makes the *unmodified* in-tree model match the reference to max-abs 7e-7.

### Expected behavior

`MiniMaxM2Config(rotary_dim=64)` should yield `rope_parameters["partial_rotary_factor"] == 0.5` so the model rotates only the trained 64 dims — matching the checkpoints' original remote-code config. Fix (following the `GPTNeoXConfig` `rotary_pct` precedent): map the legacy field in a `convert_rope_params_to_dict` override. PR incoming.

### PR DESCRIPTION
**Pull request by dajiaohuang**
[CPU CI](https://transformers-ci.lor-e.huggingface.cool/d/pytest-observability-by-pr/pytest-observability-branch?var-pr=48486) [GPU run-slow](https://transformers-ci.lor-e.huggingface.cool/d/pytest-observability-by-pr/pytest-observability-branch?var-pr=48486)

## What does this PR do?

Fixes #48241.

Released MiniMaxAI/MiniMax-M2* checkpoints are trained with partial RoPE (`rotary_dim: 64` of `head_dim: 128`) and express it through a legacy `rotary_dim` config field; their original remote-code config derived `partial_rotary_factor = rotary_dim / head_dim` in `__init__`. The in-tree `MiniMaxM2Config` keeps `rotary_dim` as an opaque extra attribute, so `rope_parameters` never receives a partial factor and `MiniMaxM2RotaryEmbedding` silently rotates the full head dimension with a full-width frequency ladder — wrong frequency spacing *and* wrong rotate-half pairing in every attention layer (measured: mean KL 13.1 / cosine 0.02 vs a faithful reference on MiniMax-M2.7 over 2,048 tokens; details in the issue).

This maps the legacy field in a `convert_rope_params_to_dict` override on `MiniMaxM2Config` (in `modular_minimax_m2.py`), following the `GPTNeoXConfig` `rotary_pct` precedent:

- `MiniMaxM2Config(rotary_dim=64)` → `rope_parameters["partial_rotary_factor"] == 0.5`, `inv_freq` length 32;
- an explicit `partial_rotary_factor` in `rope_parameters` still wins;
- configs without `rotary_dim` are unaffected.

## Before submitting

- [x] This PR fixes a typo or improves the docs
- [x] Did you read the contributor guideline?
- [x] Was this discussed/approved via a Github issue? #48241
- [x] Did you make sure to update the documentation with your changes? N/A
- [x] Did you write any new necessary tests? `test_legacy_rotary_dim_maps_to_partial_rotary_factor` (covers legacy mapping, explicit-factor priority, and the no-legacy-field case)

## Who can review?

@ArthurZucker @Cyrilvallez

### CODE REVIEW

**Comment by dajiaohuang** [comment]
Fixed the missed modular conversion output in commit 26c0e74. The PR's new convert_rope_params_to_dict method is now present in generated configuration_minimax_m2.py, resolving the repository consistency diff and the KeyError in test_legacy_rotary_dim_maps_to_partial_rotary_factor. Validation: Python compileall passed and git diff --check passed.

**Comment by dajiaohuang** [comment]
Follow-up after the latest consistency run: the failure also compared against upstream main at c119ec3cc3, which had advanced beyond the fork's stale main and introduced inherited conversion output in minimax_m3_vl/step3p7. I rebased the existing branch onto current upstream main and regenerated the MiniMax M2 config output, including exact converter formatting, in commit e434db8. The branch now contains only the modular source, matching generated configuration, and focused test changes. The earlier CI failure was reproduced from job 100931027864.

**Comment by dajiaohuang** [comment]
The latest consistency failure (run 33845763052 / job 100937312164) was from generated descendants of MiniMaxM2Config: minimax_m3_vl and step3p7 were missing the inherited converter override, while the MiniMaxM2 generated file itself was correct. Regenerated both configuration files on top of upstream main and pushed as a20b527110. The focused modular conversion check now passes with UTF-8 enabled.

**Comment by YeonwooSung** [comment]
Not trying to widen this PR — just a descendant check.

Official `MiniMaxAI/MiniMax-M3` / `MiniMax-M3-MXFP8` already ship `text_config.partial_rotary_factor: 0.5` next to `rotary_dim: 64`, so `from_pretrained` on those repos is already on the 0.5 path. Official M3 checkpoints are fine.

The remaining hole is only a bare `MiniMaxM3VLTextConfig()` / a `rotary_dim`-only hand-built config. The M3 tests set both fields, so they do not catch that path.

`MiniMaxM3VLTextConfig` subclasses `MiniMaxM2Config` and defaults `rotary_dim=64`. If this mapping lands on M2, `make fix-repo` will copy it onto the generated M3-VL config and the default M3-VL config becomes 64/128 = 0.5. That matches the M3 docstring, but it is a default-value change versus today's implicit 1.0.

Is that inheritance intended here, or do you want M3-VL left out of this PR?

**Comment by dajiaohuang** [comment]
Yes, M3-VL is intentionally covered. MiniMaxM3VLTextConfig documents rotary_dim=64 as the number of channels rotated and inherits the M2 configuration, so a bare config should resolve to partial_rotary_factor=0.5; released M3 configs already carry the same 64/128 setting. The generated descendant override keeps that inherited behavior consistent. I added a regression covering the bare default and explicit-factor precedence in a96d37c (compileall and git diff --check pass). The focused pytest is currently blocked by this checkout environment: tokenizers==0.22.2, while this Transformers revision requires >=0.23.1 and <0.24.0.

**Review by Cyrilvallez — APPROVED** [review]
Allright thanks, I fixed all that was wrong - LGTM now

**Comment by vasqu** [comment]
That's actually on the minimax team, they never came around to merge the PRs on the hub... e.g. https://huggingface.co/MiniMaxAI/MiniMax-M2/discussions/52
